### PR TITLE
refactor for "portable tags"

### PR DESCRIPTION
**Pull request by snopoke**
## Summary
I've been in this code a bit recently looking at some timezone display issue.

This code has always bothered me because it is very hard to understand. This is mostly because it's driven by dynamic configuration inputs which were vanilla dicts. In addition to that the configs could come from CouchDB making impossible to know the full set of possible configs.

In addition to some basic refactoring this PR changes that in the following ways:
* configs are no longer dicts, they are data classes
* configs can no longer come from the DB (see comment on cfed00c)
* only support a fixed set of processors
* convert date parsing to a processor instead of a separate option

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
This did have some automated coverage but I have added more.

### QA Plan
QA on staging

### Safety story
This has the potential to impact the case data and form data reports in addition to one custom report. It does not affect any backend processing and it should not impact the UI at all since it is only a refactor.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Review by biyeun — APPROVED** [review]
this looks great! a huge improvement tbh 🎉

File: corehq/apps/hqwebapp/templatetags/proptable_tags.py [hunk 1]
```diff
@@ -113,60 +132,76 @@ def _key_format(k, v):
...
 
 
-def get_display_data(data, prop_def, processors=None, timezone=pytz.utc):
-    # when prop_def came from a couchdbkit document, it will be a LazyDict with
-    # a broken pop method.  This conversion also has the effect of a shallow
-    # copy, which we want.
-    prop_def = dict(prop_def)
+@attr.s
+class DisplayConfig:
+    # dict key for callable to get value from data dict
```
Comment (snopoke) [inline corehq/apps/hqwebapp/templatetags/proptable_tags.py:137]:
```suggestion
    # dict key or callable to get value from data dict
```

File: corehq/apps/hqwebapp/templatetags/proptable_tags.py [hunk 2]
```diff
@@ -113,60 +132,76 @@ def _key_format(k, v):
...
 
-    default_processors = {
-        'yesno': yesno,
-        'doc_info': lambda value: pretty_doc_info(
-            get_doc_info_by_id(data['domain'], value)
-        )
-    }
-    processors = processors or {}
-    processors.update(default_processors)
+    # name of the field. Defaults to `expr` if not given.
```
Comment (snopoke) [inline corehq/apps/hqwebapp/templatetags/proptable_tags.py:140]:
```suggestion
    # name of the field. Defaults to the value `expr` if not given.
```

**Comment by snopoke** [comment]
QA ticket: https://dimagi-dev.atlassian.net/browse/QA-2816

**Pull request merged**